### PR TITLE
Remove duplicate stats aggregator from V2 query metrics

### PR DESCRIPTION
**Pull request by KKcorps**
- Fix the TODO in the V2 metrics where we create a duplicate class to handle stats. 
- The PR refactors the existing aggregator class to handle both V1 and V2 metrics.
- **Bug Fix**: Add the server name to the metric key so that metrics from multiple servers do not replace each other.
- Also, add tests for V2 metric.

### CODE REVIEW

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxReceiveOperator.java [hunk 1]
```diff
@@ -164,7 +164,6 @@ protected TransferableBlock getNextBlock() {
               return block;
             } else {
               if (!block.getResultMetadata().isEmpty()) {
-                _operatorStats.clearExecutionStats();
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxReceiveOperator.java:167]: why is this necessary previously? i thought the stats aggregator is only created once right?

File: pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java [hunk 2]
```diff
@@ -56,7 +56,7 @@ public void testThreadExceptionTransfer() {
               threadPoolService,
               ImmutableMap.of(routingInstance, mockedResponse),
               1000,
-              mock(BaseReduceService.ExecutionStatsAggregator.class));
+              mock(ExecutionStatsAggregator.class));
```
Comment (walterddr) [inline pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java:59]: could we add some test for both v1 and v2 engine for collecting and aggregating stats via different metadata blocks?

**Review by walterddr — COMMENTED** [review]
nice refactoring!

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorStats.java [hunk 3]
```diff
@@ -61,11 +61,11 @@ public void recordRow(int numBlock, int numRows) {
   }
 
   public void recordExecutionStats(Map<String, String> executionStats) {
-    _statsAggregator.aggregate(executionStats);
+    _executionStats = executionStats;
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorStats.java:64]: could we think of a way to incorporate `_numBlock` and `_numRows` in the executionStats?

Comment (KKcorps) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorStats.java:64]: Makes sense. I was thinking about that as well.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxReceiveOperator.java (hunk 1 above)
Comment (KKcorps) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxReceiveOperator.java:167]: Previously we were aggregating stats, hence there was need to clear them. Now, we are just replacing the stats.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorStats.java [hunk 4]
```diff
@@ -61,11 +63,14 @@ public void recordRow(int numBlock, int numRows) {
...
   }
 
   public Map<String, String> getExecutionStats() {
-    return _statsAggregator.getStats();
+    _executionStats.put(OperatorUtils.NUM_BLOCKS, String.valueOf(_numBlock));
+    _executionStats.put(OperatorUtils.NUM_ROWS, String.valueOf(_numRows));
+    _executionStats.put(OperatorUtils.WALL_TIME, String.valueOf(_executeStopwatch.elapsed(TimeUnit.MILLISECONDS)));
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorStats.java:72]: this is technically thread-time not WALL_TIME. let's change the naming

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java [hunk 5]
```diff
@@ -83,14 +84,22 @@ public class AggregateOperator extends MultiStageOperator {
   public AggregateOperator(MultiStageOperator inputOperator, DataSchema dataSchema, List<RexExpression> aggCalls,
       List<RexExpression> groupSet, DataSchema inputSchema, long requestId, int stageId) {
     this(inputOperator, dataSchema, aggCalls, groupSet, inputSchema, AggregateOperator.Accumulator.MERGERS, requestId,
-        stageId);
+        stageId, null);
+  }
+
+  public AggregateOperator(MultiStageOperator inputOperator, DataSchema dataSchema, List<RexExpression> aggCalls,
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:90]: can we replace the constructor by adding virtual server instead of creating one? which function is still using the constructor without the virtual server address?

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTestBase.java [hunk 6]
```diff
@@ -111,6 +112,37 @@ protected List<Object[]> queryRunner(String sql) {
         queryPlan.getQueryResultFields(), queryPlan.getQueryStageMap().get(0).getDataSchema()).getRows();
   }
 
+  protected List<Object[]> queryRunner(String sql, ExecutionStatsAggregator executionStatsAggregator) {
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTestBase.java:115]: simply make the normal queryRunner API call this one with a null executionStatsAggregator?

the only difference is the last step when reduceMailboxReceive yes?

File: pinot-query-runtime/src/test/resources/metadata_queries/BasicQuery.json [hunk 7]
```diff
@@ -0,0 +1,68 @@
+{
```
Comment (walterddr) [inline pinot-query-runtime/src/test/resources/metadata_queries/BasicQuery.json:1]: why are we duplicating the ResourceBasedTest?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java (hunk 5 above)
Comment (KKcorps) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:90]: Mostly test classes

Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:90]: oh. i see. in this case lets make it clean -- there's no need to keep additional public constructor just for test purposes.

Comment (KKcorps) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java]: Done.

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTestBase.java (hunk 6 above)
Comment (KKcorps) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTestBase.java:85]: Done.

**Review by walterddr — APPROVED** [review]

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/ResourceBasedQueriesTest.java [hunk 8]
```diff
@@ -260,6 +294,48 @@ private static Object[][] testResourceQueryTestCaseProviderBoth()
...
+      for (QueryTestCase.Query queryCase : queryCases) {
+        if (queryCase._ignored) {
+          continue;
+        }
+
+        if (queryCase._outputs != null) {
+          String sql = replaceTableName(testCaseName, queryCase._sql);
+          if (!sql.contains("basic_test")) {
+            continue;
+          }
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/ResourceBasedQueriesTest.java:318]: technically you can test this stats with numOfSegments on all test cases. also you can directly add a test to QueryRunnerTest instead of doing it here. but I will leave this up to you.. 

my thinking is: since you are not testing many different scenarios. it is best to just have 1 `@Test` section in QueryRunnerTest

Comment (KKcorps) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/ResourceBasedQueriesTest.java:318]: Can't do for all test cases since the metrics contain only segments queries from the servers but doesn't account for the segments that are pruned by the brokers. So in some cases the test fails.
